lib/Scheduling/Scheduler.cpp

```diff
@@ -0,0 +1,156 @@
+//===- Scheduler.cpp - Common interface for scheduling algorithms -===========//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements base classes and utilities for scheduling algorithms.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Scheduling/Scheduler.h"
+
+#include "mlir/IR/Operation.h"
+
+using namespace circt;
+using namespace circt::sched;
+
+Scheduler::Scheduler(Operation *containingOp) : containingOp(containingOp) {
+  uniquer.registerParametricStorageType<Dependence>();
+  uniquer.registerParametricStorageType<OperatorType>();
+}
+
+Dependence *Scheduler::getOrInsertDependence(Operation *src, unsigned srcIdx,
+                                             Operation *dst, unsigned dstIdx) {
+  auto *dep = uniquer.get<Dependence>({}, src, dst, 0, 0);
+  deps.insert(dep);
+  return dep;
+}
+
+OperatorType *Scheduler::getOrInsertOperatorType(StringRef name) {
+  auto *opr = uniquer.get<OperatorType>({}, name);
+  oprs.insert(opr);
+  return opr;
+}
+
+LogicalResult Scheduler::checkOp(Operation *op) {
+  if (!hasAssociatedOperatorType(op)) {
+    op->emitError("Operation is not associated with an operator type");
```

Comment:
FYI you can `return op->emitError(...)` in failure cases. This will also make this a single line `if` so you could remove the curly braces. This comment applies to several similar cases below.

Reply:
Nice, didn't know that. Changed.